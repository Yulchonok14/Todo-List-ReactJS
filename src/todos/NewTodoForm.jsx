import React, { useState } from 'react';

import styled from 'styled-components';

const NewTodoFormStyle = styled.div`
    border-radius: 8px;
    padding: 16px;
    text-align: center;
    box-shadow: 0 4px 8px grey;
`;

const NewTodoInput = styled.input`
    font-size: 16px;
    padding: 8px;
    border: none;
    border-bottom: 2px solid #ddd;
    border-radius: 8px;
    width: 70%;
    outline: none;
`;

const NewTodoButton = styled.button`
    font-size: 16px;
    padding: 8px;
    border: none;
    border-radius: 8px;
    outline: none;
    cursor: pointer;
    margin-left: 8px;
    width: 20%;
    background-color: #22ee22;
`;

const NewTodoForm = ({ todos, onCreateNewTodoClicked }) => {
  const [newTodo, setNewTodo] = useState('');

  return (
    <NewTodoFormStyle>
      <NewTodoInput
        placeholder="Type new todo item here"
        value={newTodo}
        onChange={(e) => setNewTodo(e.target.value)}
      />
      <NewTodoButton
        onClick={() => {
          const isDuplicated = todos.some((todo) => todo.text === newTodo);
          if (!isDuplicated) {
            onCreateNewTodoClicked(newTodo);
            setNewTodo('');
          }
        }}
      >
        Craft Todo
      </NewTodoButton>
    </NewTodoFormStyle>
  );
};

export default NewTodoForm;
