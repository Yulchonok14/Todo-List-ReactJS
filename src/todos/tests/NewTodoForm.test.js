/* eslint-disable react/jsx-filename-extension */
/* eslint-env mocha */
import 'jsdom-global/register';
import chai, { expect } from 'chai';
import React from 'react';
import chaiJestSnapshot from 'chai-jest-snapshot';
import renderer from 'react-test-renderer';
import { configure, mount } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import sinon from 'sinon';
import NewTodoForm from '../NewTodoForm';

chai.use(chaiJestSnapshot);

configure({
  adapter: new Adapter(),
});

before(() => {
  chaiJestSnapshot.resetSnapshotRegistry();
});

// eslint-disable-next-line func-names
beforeEach(function () {
  chaiJestSnapshot.configureUsingMochaContext(this);
});

describe('NewTodoForm', () => {
  it('renders correctly', () => {
    const component = renderer.create(<NewTodoForm />);
    const tree = component.toJSON();

    expect(tree).to.matchSnapshot();
  });

  it('has input with placeholder and button', () => {
    const mockOnCreateNewTodoCallBack = sinon.stub();
    const mockTodos = [{
      id: 1,
      text: 'go to the park',
      isCompleted: true,
    }];
    const wrapper = mount(<NewTodoForm
      onCreateNewTodoClicked={mockOnCreateNewTodoCallBack}
      todos={mockTodos}
    />);
    expect(wrapper.find('button').at(0).text().includes('Craft Todo')).to.equal(true);
    expect(wrapper.find('input').at(0).props().placeholder).to.equal('Type new todo item here');
  });

  it('invokes a callback with value if text is not duplicated', () => {
    const mockOnCreateNewTodoCallBack = sinon.stub();
    const mockTodos = [{
      id: 1,
      text: 'Go to the park',
      isCompleted: true,
    }];
    const wrapper = mount(<NewTodoForm
      onCreateNewTodoClicked={mockOnCreateNewTodoCallBack}
      todos={mockTodos}
    />);
    wrapper.find('input').simulate('change', { target: { value: 'Read a book' } });
    expect(wrapper.find('input').instance().value).to.equal('Read a book');

    wrapper.find('button').at(0).simulate('click');
    expect(mockOnCreateNewTodoCallBack.calledOnce).to.equal(true);
    expect(mockOnCreateNewTodoCallBack.calledWith('Read a book')).to.equal(true);

    expect(wrapper.find('input').instance().value).to.equal('');
  });

  it('does not invoke a callback if text is duplicated', () => {
    const mockOnCreateNewTodoCallBack = sinon.stub();
    const mockTodos = [{
      id: 1,
      text: 'Go to the park',
      isCompleted: true,
    }];
    const wrapper = mount(<NewTodoForm
      onCreateNewTodoClicked={mockOnCreateNewTodoCallBack}
      todos={mockTodos}
    />);
    wrapper.find('input').simulate('change', { target: { value: 'Go to the park' } });
    expect(wrapper.find('input').instance().value).to.equal('Go to the park');

    wrapper.find('button').at(0).simulate('click');
    expect(mockOnCreateNewTodoCallBack.calledOnce).to.equal(false);

    expect(wrapper.find('input').instance().value).to.equal('Go to the park');
  });
});
