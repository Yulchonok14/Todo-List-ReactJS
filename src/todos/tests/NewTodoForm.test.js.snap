// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewTodoForm renders correctly 1`] = `
<div
  className="sc-bdfBQB hpKxwM"
>
  <input
    className="sc-gsTEea NRtLr"
    onChange={[Function]}
    placeholder="Type new todo item here"
    value=""
  />
  <button
    className="sc-dlfnuX eOyWug"
    onClick={[Function]}
  >
    Craft Todo
  </button>
</div>
`;
